pkg/relayer/miner/miner.go

```diff
@@ -0,0 +1,408 @@
+// method to each relay. It returns an observable of the mined relays.
+func (mnr *miner) mineRelays(
+	ctx context.Context,
+	servedRelays observable.Observable[servicetypes.Relay],
+) observable.Observable[either.Either[minedRelay]] {
+	// servicetypes.Relay ==> either.Either[minedRelay]
+	return channel.Map(ctx, servedRelays, mnr.mapMineRelay)
+}
+
+// mapMineRelay is intended to be used as a MapFn. It hashes the relay and compares
+// its difficulty to the minimum threshold. If the relay difficulty is sifficient,
+// it returns an either populated with the minedRelay value. Otherwise, it skips
+// the relay. If it encounters an error, it returns an either populated with the
+// error.
+func (mnr *miner) mapMineRelay(
+	_ context.Context,
+	relay servicetypes.Relay,
+) (_ either.Either[minedRelay], skip bool) {
+	relayBz, err := relay.Marshal()
+	if err != nil {
+		return either.Error[minedRelay](err), true
+	}
+
+	// Is it correct that we need to hash the key while smst.Update() could do it
+	// since smst has a reference to the hasher
+	mnr.hasher.Write(relayBz)
+	relayHash := mnr.hasher.Sum(nil)
+	mnr.hasher.Reset()
+
+	if !protocol.BytesDifficultyGreaterThan(relayHash, defaultRelayDifficulty) {
+		return either.Success(minedRelay{}), true
+	}
+
+	return either.Success(minedRelay{
+		Relay: relay,
+		bytes: relayBz,
+		hash:  relayHash,
+	}), false
+}
+
+// addReplayToSessionTree maps over the eitherMinedRelays observable, applying the
+// mapAddRelayToSessionTree method to each relay. It returns an observable of the
+// errors encountered.
+func (mnr *miner) addReplayToSessionTree(
+	ctx context.Context,
+	eitherMinedRelays observable.Observable[either.Either[minedRelay]],
+) observable.Observable[error] {
+	// either.Either[minedRelay] ==> error
+	return channel.Map(ctx, eitherMinedRelays, mnr.mapAddRelayToSessionTree)
+}
+
+// mapAddRelayToSessionTree is intended to be used as a MapFn. It adds the relay
+// to the session tree. If it encounters an error, it returns the error. Otherwise,
+// it skips output (only outputs errors).
+func (mnr *miner) mapAddRelayToSessionTree(
+	_ context.Context,
+	eitherRelay either.Either[minedRelay],
+) (_ error, skip bool) {
+	// Propagate any upstream errors.
+	relay, err := eitherRelay.ValueOrError()
+	if err != nil {
+		return err, false
+	}
+
+	// ensure the session tree exists for this relay
+	sessionHeader := relay.GetReq().GetMeta().GetSessionHeader()
+	smst, err := mnr.sessionManager.EnsureSessionTree(sessionHeader)
+	if err != nil {
+		log.Printf("failed to ensure session tree: %s\n", err)
+		return err, false
+	}
+
+	if err := smst.Update(relay.hash, relay.bytes, 1); err != nil {
+		log.Printf("failed to update smt: %s\n", err)
+		return err, false
+	}
+
+	// Skip because this map function only outputs errors.
+	return nil, true
+}
+
+// waitForOpenClaimWindow maps over the SessionsToClaim observable, applying the
```

Comment:
1. Comment is incorrect `s/mapWaitForOpenClaimWindow/waitForOpenClaimWindow`
2. It does a map under the hood but the function is not called map (the one below is)
3. I personally think that having the business logic inline + comment would lead to less indirection. I'm getting a bit confused jumping around all the mapper & functions